### PR TITLE
Iterate on train loop

### PR DESCRIPTION
**Pull request by guarin**
## What has changed and why?

* Init dataloader
* Init model
* Minimal loop

## How has it been tested?

(Delete this: Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration.)

## Did you update [CHANGELOG.md](../CHANGELOG.md)?

- [ ] Yes
- [x] Not needed (internal change)

## Did you update the documentation?

- [ ] Yes
- [x] Not needed (internal change without effects for user)

### CODE REVIEW

File: src/lightly_train/_commands/common_helpers.py [hunk 1]
```diff
@@ -542,3 +542,63 @@ def _unlink_and_ignore(path: Path) -> None:
         path.unlink(missing_ok=True)
     except OSError:
         pass
+
+
+_T = TypeVar("_T")
+
+
+def get_global_batch_size(
+    global_batch_size: int,
```
Comment (guarin) [inline src/lightly_train/_commands/common_helpers.py:551]: Moved from `train_helpers.py` to `common_helpers.py`

File: src/lightly_train/_commands/train_helpers.py [hunk 2]
```diff
@@ -93,63 +93,6 @@ def get_total_num_devices(
     return total_num_devices
 
 
-def get_global_batch_size(
-    global_batch_size: int,
```
Comment (guarin) [inline src/lightly_train/_commands/train_helpers.py:97]: Moved to `common_helpers.py`

File: src/lightly_train/_commands/train_task_helpers.py [hunk 3]
```diff
@@ -96,5 +127,88 @@ def get_dataset(
...
+
+
+def get_task_train_model_args(
+    task_args: dict[str, Any] | TaskTrainModelArgs | None,
+) -> TaskTrainModelArgs:
+    if isinstance(task_args, TaskTrainModelArgs):
+        return task_args
+    task_args = {} if task_args is None else task_args
+    task_cls = DINOv2SemanticSegmentationTrainArgs
+    args = validate.pydantic_model_validate(task_cls, task_args)
```
Comment (guarin) [inline src/lightly_train/_commands/train_task_helpers.py:206]: Will have to make this more flexible once we have more tasks

File: src/lightly_train/_data/infinite_cycle_iterator.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    def __iter__(self) -> Self:
+        return self
+
+    def __next__(self) -> _T:
+        if self._iter is None:
+            self._iter = iter(self.iterable)
+        try:
+            return next(self._iter)
+        except StopIteration:
+            self._iter = iter(self.iterable)
```
Comment (stegmuel) [inline src/lightly_train/_data/infinite_cycle_iterator.py:33]: Is it correct to assume that fabric takes care of doing `self.iterable.sampler.set_epoch(epoch)` ?

Comment (stegmuel) [reply src/lightly_train/_data/infinite_cycle_iterator.py:33]: Can we have a TODO to replace this with an infinite sampler _à la_ DINOv2 as this approach still has the same overhead as the epoch-based approach.

File: src/lightly_train/_commands/train_task.py [hunk 5]
```diff
@@ -96,28 +108,83 @@ def train_task_from_config(config: TrainTaskConfig) -> None:
...
 
 
 class TrainTaskConfig(PydanticConfig):
     out: PathLike
     data: MaskSemanticSegmentationDataArgs
     model: str
     task: Literal["semantic_segmentation"]
+    steps: int | Literal["auto"] = "auto"
+    batch_size: int | Literal["auto"] = "auto"
+    num_workers: int | Literal["auto"] = "auto"
```
Comment (stegmuel) [inline src/lightly_train/_commands/train_task.py:193]: isn't this an attribute of `loader_args`?

File: src/lightly_train/_commands/train_task.py [hunk 6]
```diff
@@ -96,28 +108,83 @@ def train_task_from_config(config: TrainTaskConfig) -> None:
...
     logger.info(
-        f"Resolved Args: {train_task_helpers.pretty_format_args(args=config.model_dump())}"
+        f"Resolved Args: {helpers.pretty_format_args(args=config.model_dump())}"
     )
+    logger.info(f"Starting training for {config.steps} steps...")
+    for step in range(config.steps):
+        batch = next(infinite_train_dataloader)
+        # TODO(Guarin, 07/25): Backprop.
+        # TODO(Guarin, 07/25): Log loss and metrics.
+        model.training_step(fabric=fabric, batch=batch)
```
Comment (stegmuel) [inline src/lightly_train/_commands/train_task.py:174]: `loss.backward()` and `optimizer.step()` are also implemented inside `model.trainig_step`, right?

**Review by stegmuel — APPROVED** [review]
Approved with some questions.

File: src/lightly_train/_data/infinite_cycle_iterator.py (hunk 4 above)
Comment (guarin) [reply src/lightly_train/_data/infinite_cycle_iterator.py:33]: > Is it correct to assume that fabric takes care of doing self.iterable.sampler.set_epoch(epoch) ?

Yes this happens here:  https://github.com/Lightning-AI/pytorch-lightning/blob/e29a8cbb83b9041ff7b7d10f24d9274f39d2358d/src/lightning/fabric/wrappers.py#L320-L321

> Can we have a TODO to replace this with an infinite sampler à la DINOv2 as this approach still has the same overhead as the epoch-based approach.

Added

File: src/lightly_train/_commands/train_task.py (hunk 5 above)
Comment (guarin) [reply src/lightly_train/_commands/train_task.py:193]: It is the same as in `train.py`. `batch_size` and `num_workers` are important and often changed by users so we expose them directly. Rarely used arguments like `persistent_workers` are not directly exposed but can be set trough `loader_args`.

File: src/lightly_train/_commands/train_task.py (hunk 6 above)
Comment (guarin) [reply src/lightly_train/_commands/train_task.py:174]: I would have put it outside together with `optimizer.zero_grad`, gradient clipping, gradient accumulation, etc. Otherwise we have to reimplement it in every model.

Comment (stegmuel) [reply src/lightly_train/_commands/train_task.py:174]: So do I, but I was confused to not see it.

Comment (guarin) [reply src/lightly_train/_commands/train_task.py:174]: Couldn't implement it yet because the forward pass is not implemented yet :)
